File: onlyskills-repo/repo_scanner.pl
% Repo Scanner - Prolog
:- module(repo_scanner, [
    scan_repos/2,
    find_duplicates/2,
    prove_novelty/2
]).

:- use_module(library(filesex)).

% Scan all repos
scan_repos(BaseDir, Scans) :-
    find_git_repos(BaseDir, Repos),
    maplist(scan_repo, Repos, Scans).

% Find git repos
find_git_repos(BaseDir, Repos) :-
    directory_files(BaseDir, Files),
    findall(Repo,
        (member(File, Files),
         atom_concat(BaseDir, '/', Base),
         atom_concat(Base, File, Repo),
         exists_directory(Repo),
         atom_concat(Repo, '/.git', GitDir),
         exists_directory(GitDir)),
        Repos).

% Scan single repo
scan_repo(Repo, Scan) :-
    file_base_name(Repo, Name),
    find_docs(Repo, Docs),
    length(Docs, DocCount),
    hash_to_shard(Name, Shard),
    Scan = [
        repo(Repo),
        name(Name),
        shard(Shard),
        doc_count(DocCount)
    ].

% Find docs
find_docs(Repo, Docs) :-
    findall(Doc,
        (directory_member(Repo, Doc, [recursive(true)]),
         file_name_extension(_, Ext, Doc),
         member(Ext, [md, txt, rst])),
        Docs).

% Hash to shard
hash_to_shard(Name, Shard) :-
    atom_codes(Name, Codes),
    sum_list(Codes, Sum),
    Shard is Sum mod 71.

% Find duplicates in 8M files
find_duplicates(Scans, Duplicates) :-
    findall(Name-Count,
        (member(Scan, Scans),
         member(name(Name), Scan),
         findall(S, (member(S, Scans), member(name(Name), S)), Matches),
         length(Matches, Count),
         Count > 1),
        Duplicates).

% Prove novelty (no duplicates)
prove_novelty(Scans, Proof) :-
    find_duplicates(Scans, Duplicates),
    (Duplicates = [] ->
        Proof = [novel(true), duplicates(0)]
    ; length(Duplicates, Count),
      Proof = [novel(false), duplicates(Count)]).

File: onlyskills-repo/src/bin/repo_scanner.rs
// Repo Scanner - Rust (replaces scan_all_repos.py)
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::collections::HashMap;
use walkdir::WalkDir;

#[derive(Debug, Clone, Serialize)]
struct RepoScan {
    repo: PathBuf,
    name: String,
    shard: u8,
    doc_count: usize,
    docs: Vec<PathBuf>,
}

fn hash_to_shard(data: &str) -> u8 {
    (data.bytes().map(|b| b as u32).sum::<u32>() % 71) as u8
}

fn find_all_repos(base: &Path) -> Vec<PathBuf> {
    WalkDir::new(base)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_name() == ".git" && e.file_type().is_dir())
        .filter_map(|e| e.path().parent().map(|p| p.to_path_buf()))
        .collect()
}

fn find_docs(repo: &Path) -> Vec<PathBuf> {
    WalkDir::new(repo)
        .max_depth(3)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_type().is_file() && 
            e.path().extension().map_or(false, |ext| 
                ext == "md" || ext == "txt" || ext == "rst"
            )
        })
        .map(|e| e.path().to_path_buf())
        .collect()
}

fn scan_repo(repo: &Path) -> RepoScan {
    let name = repo.file_name().unwrap().to_string_lossy().to_string();
    let docs = find_docs(repo);
    let shard = hash_to_shard(&name);
    
    RepoScan {
        repo: repo.to_path_buf(),
        name,
        shard,
        doc_count: docs.len(),
        docs: docs.into_iter().take(10).collect(),
    }
}

fn main() {
    println!("🔍 Repo Scanner (Rust)");
    println!("{}", "=".repeat(70));
    
    let base = Path::new("/home/mdupont");
    let repos = find_all_repos(base);
    
    println!("Found {} repos", repos.len());
    
    let scanned: Vec<_> = repos.iter().take(100)
        .map(|r| scan_repo(r))
        .collect();
    
    let mut by_shard: HashMap<u8, Vec<&RepoScan>> = HashMap::new();
    for scan in &scanned {
        by_shard.entry(scan.shard).or_default().push(scan);
    }
    
    println!("Distributed across {} shards", by_shard.len());
    
    let json = serde_json::to_string_pretty(&scanned).unwrap();
    std::fs::write("repo_scan_rust.json", json).unwrap();
    
    println!("✓ Saved: repo_scan_rust.json");
}
